src/control/ros/Conversions.cpp

```diff
@@ -0,0 +1,322 @@
+    }
+  }
+
+  return joints;
+}
+
+} // namespace
+
+//=============================================================================
+std::unique_ptr<SplineTrajectory> convertJointTrajectory(
+  const std::shared_ptr<MetaSkeletonStateSpace>& space,
+  const trajectory_msgs::JointTrajectory& jointTrajectory)
+{
+  if (!space)
+    throw std::invalid_argument{"StateSpace must be non-null."};
+
+  const auto numControlledJoints = space->getNumSubspaces();
+  if (jointTrajectory.joint_names.size() != numControlledJoints)
+  {
+    std::stringstream message;
+    message << "Incorrect number of joints: expected "
+        << numControlledJoints << ", got "
+        << jointTrajectory.joint_names.size() << ".";
+    throw std::invalid_argument{message.str()};
+  }
+
+  // Check that the names in jointTrajectory are unique.
+  std::vector<std::string> joint_names;
+  joint_names.reserve(numControlledJoints);
+  for (size_t i = 0; i < numControlledJoints; ++i)
+  {
+    joint_names.emplace_back(jointTrajectory.joint_names[i]);
+  }
+  std::sort(joint_names.begin(), joint_names.end());
+  for (size_t i = 0; i < numControlledJoints - 1; ++i)
+  {
+    if (joint_names[i] == joint_names[i+1])
+    {
+      std::stringstream message;
+      message << "JointTrajectory has multiple joints with same name ["
+        << joint_names[i] << "].";
+      throw std::invalid_argument{message.str()};
+    }
+  }
+
+  // Check that all joints are single-DOF RnJoint or SO2JOint state spaces.
+  for (size_t i = 0; i < space->getNumSubspaces(); ++i)
+  {
+    auto joint = space->getJointSpace(i)->getJoint();
+    auto jointSpace = space->getSubspace(i);
+    auto rnJoint = dynamic_cast<RnJoint*>(jointSpace.get());
+    auto so2Joint = dynamic_cast<SO2Joint*>(jointSpace.get());
+
+    if (joint->getNumDofs() != 1 || (!rnJoint && !so2Joint))
+    {
+      std::stringstream message;
+      message << "Only single-DOF RnJoint and SO2Joint are supported. Joint "
+        << joint->getName() << "(index: " << i << ") is a "
+        << joint->getType() << " with "
+        << joint->getNumDofs() << " DOFs.";
+      throw std::invalid_argument{message.str()};
+    }
+  }
+
+  if (jointTrajectory.points.size() < 2)
+  {
+    throw std::invalid_argument{
+      "Trajectory must contain two or more waypoints."};
+  }
+
+  // Map joint indices between jointTrajectory and space subspaces.
+  std::map<size_t, size_t> rosJointToMetaSkeletonJoint;
+
+  auto metaSkeleton = space->getMetaSkeleton();
+  auto nJoints = jointTrajectory.joint_names.size();
+
+  for (size_t trajIndex = 0; trajIndex < nJoints; ++trajIndex)
+  {
+    const auto& jointName = jointTrajectory.joint_names[trajIndex];
+    auto joints = findJointByName(*metaSkeleton, jointName);
+
+    if (joints.size() == 0)
+    {
+      std::stringstream message;
+      message << "Joint " << jointName << " (index: " << trajIndex << ")"
+        << " does not exist in metaSkeleton.";
```

Comment:
Nit: This index is ambiguous because it could refer to an index in the `JointTrajectory` or an index in the `MetaSkeleton`. I suggest explicitly saying "trajectory index."